### PR TITLE
ENH: to_datetime support iso week year (16607)

### LINKED ISSUE #16607: to_datetime should support ISO week year
`to_datetime` does not currently seem to support `ISO week year` like `strptime` does:

```
In [38]: datetime.date(2016, 1, 1).strftime('%G-%V')
Out[38]: '2015-53'

In [39]: datetime.datetime.strptime(datetime.date(2016, 1, 1).strftime('%G-%V')+'-1', '%G-%V-%u')
Out[39]: datetime.datetime(2015, 12, 28, 0, 0)

In [41]: pd.to_datetime(datetime.date(2016, 1, 1).strftime('%G-%V')+'-1', format='%G-%V-%u')
        ---------------------------------------------------------------------------
        TypeError                                 Traceback (most recent call last)
        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in _convert_listlike(arg, box, format, name, tz)
            443             try:
        --> 444                 values, tz = tslib.datetime_to_datetime64(arg)
            445                 return DatetimeIndex._simple_new(values, name=name, tz=tz)

        pandas/_libs/tslib.pyx in pandas._libs.tslib.datetime_to_datetime64 (pandas/_libs/tslib.c:33275)()

        TypeError: Unrecognized value type: <class 'str'>

        During handling of the above exception, another exception occurred:

        ValueError                                Traceback (most recent call last)
        <ipython-input-41-7ce30c959690> in <module>()
        ----> 1 pd.to_datetime(datetime.date(2016, 1, 1).strftime('%G-%V')+'-1', format='%G-%V-%u')

        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in to_datetime(arg, errors, dayfirst, yearfirst, utc, box, format, exact, unit, infer_datetime_format, origin)
            516         result = _convert_listlike(arg, box, format)
            517     else:
        --> 518         result = _convert_listlike(np.array([arg]), box, format)[0]
            519 
            520     return result

        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in _convert_listlike(arg, box, format, name, tz)
            445                 return DatetimeIndex._simple_new(values, name=name, tz=tz)
            446             except (ValueError, TypeError):
        --> 447                 raise e
            448 
            449     if arg is None:

        /Users/Robin/.pyenv/versions/3.6.1/lib/python3.6/site-packages/pandas/core/tools/datetimes.py in _convert_listlike(arg, box, format, name, tz)
            412                     try:
            413                         result = tslib.array_strptime(arg, format, exact=exact,
        --> 414                                                       errors=errors)
            415                     except tslib.OutOfBoundsDatetime:
            416                         if errors == 'raise':

        pandas/_libs/tslib.pyx in pandas._libs.tslib.array_strptime (pandas/_libs/tslib.c:63124)()

        pandas/_libs/tslib.pyx in pandas._libs.tslib.array_strptime (pandas/_libs/tslib.c:63003)()

        ValueError: 'G' is a bad directive in format '%G-%V-%u'

```

<details>
INSTALLED VERSIONS
------------------
commit: None

pandas: 0.20.1
pytest: 3.1.0
pip: 9.0.1
setuptools: 28.8.0
Cython: 0.25.2
numpy: 1.12.1
scipy: 0.19.0
xarray: None
IPython: 6.0.0
sphinx: None
patsy: 0.4.1
dateutil: 2.6.0
pytz: 2017.2
blosc: None
bottleneck: None
tables: 3.4.2
numexpr: 2.6.2
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999999999
sqlalchemy: 1.1.10
pymysql: None
psycopg2: 2.7.1 (dt dec pq3 ext lo64)
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by RjLi13**
- [x] closes #16607
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I messed up the rebase of the old PR #24844 and started this again on a clean state. Reviewers please review on this PR! 

#### From old PR

I found the issue stagnant, but since the fix was already in place, I manually took the code @rosygupta made and applied it onto latest master. Rebase wasn't an option that I found would work since the file in question has been split into multiple. Let me know what else needs to be updated.

### CODE REVIEW

**Comment by RjLi13** [comment]
Just some things from the previous PR. Up to date with master. Moved to 0.25 with up to date description. Added docstrings to the function calc_julian.

#### Things needed to be addressed
- Locale tests not passing on Linux on Python 3.x environments with a locale override
  - zh_CN.UTF-8
  - it_IT.UTF-8
  - I need some help figuring out where the issue lies. It seems to be triggering the wrong error message @gfyoung 
- @jreback on splitting the test_tools testcase? I didn't quite understand the conversation.
- Other comments made by @jreback 

@gfyoung @jreback @WillAyd

**Comment by gfyoung** [comment]
@RjLi13 : With the branch rebased, are you now by any chance able to reproduce any of the test failures locally using the override methods that we discussed in the old PR?

**Comment by RjLi13** [comment]
@gfyoung Yes I get the same error as what the CI checks, using the `zh_CN.UTF-8` locale override.

**Comment by gfyoung** [comment]
Okay, that's good.  At least you can debug locally this time around.

**Comment by RjLi13** [comment]
The code was not passing the CI checks for the locale builds because the regex could not match the correct string with the pattern generated from the passed format, and it threw a different error ending the code before it got to the correct error. See this code here: https://github.com/pandas-dev/pandas/blob/947c346e860e88134f4b5032f410265f13f7a0ad/pandas/_libs/tslibs/strptime.pyx#L152-L158 
My tests for failures were not expecting to have that error thrown (see test_tools.py changes)

As for why the regex could not match the pattern, my guess is that the locale could not detect English, given the string to search for was '1999 Monday' and the regex pattern for 'zh_CN.utf8' was `re.compile('(?P<G>\\d\\d\\d\\d)\\s+(?P<A>星期一|星期二|星期三|星期四|星期五|星期六|星期日)', re.IGNORECASE)`.

**Comment by RjLi13** [comment]
As for the fix, I'm not quite sure what to do. The code works correctly, independent of the locale, because the format wanted for iso_time is %G-%V-%u, which given this snippet https://github.com/pandas-dev/pandas/blob/947c346e860e88134f4b5032f410265f13f7a0ad/pandas/_libs/tslibs/strptime.pyx#L546-L574

shows the regex is static. 

The iso code references the strptime function written for cpython [here](https://github.com/python/cpython/blob/master/Lib/_strptime.py) quite liberally.

My solution was essentially to check for the locales currently being overridden, and not run the ValueError tests for them. Alternative solutions include getting rid of checking the message matches for the ValueError, since both throw ValueErrors, or deleting the tests that use a format dependent on locale. I'm not sure how to proceed, so let me know!

File: pandas/_libs/tslibs/strptime.pyx [hunk 1]
```diff
@@ -597,7 +636,16 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
                               int day_of_week, int week_starts_Mon):
     """Calculate the Julian day based on the year, week of the year, and day of
     the week, with week_start_day representing whether the week of the year
-    assumes the week starts on Sunday or Monday (6 or 0)."""
+    assumes the week starts on Sunday or Monday (6 or 0).
+
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: please use numpydoc format here

File: pandas/_libs/tslibs/strptime.pyx [hunk 2]
```diff
@@ -620,6 +668,32 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
         return 1 + days_to_week + day_of_week
 
 
+cdef _calc_julian_from_V(int iso_year, int iso_week, int iso_weekday):
+    """Calculate the Julian day based on the ISO 8601 year, week, and weekday.
+    ISO weeks start on Mondays, with week 01 being the week containing 4 Jan.
+    ISO week days range from 1 (Monday) to 7 (Sunday).
+
+    :param iso_year: the year taken from format %G
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: same

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 3]
```diff
@@ -247,6 +247,57 @@ def test_to_datetime_parse_timezone_keeps_name(self):
...
+         "'%Y'. Use the ISO year '%G' instead.", "1999 51 Sun", "%Y %V %a"],
+        ["ISO week directive '%V' is incompatible with the year directive "
+         "'%Y'. Use the ISO year '%G' instead.", "1999 51 1", "%Y %V %w"],
+        ["ISO week directive '%V' is incompatible with the year directive "
+         "'%Y'. Use the ISO year '%G' instead.", "1999 51 1", "%Y %V %u"],
+        ["ISO week directive '%V' must be used with the ISO year directive "
+         "'%G' and a weekday directive '%A', '%a', '%w', or '%u'.", "20", "%V"]
+    ])
+    def test_ValueError_iso_week_year(self, msg, s, _format):
+        # See GH#16607
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: can you put a comment here on why why you are allowing *some* here.

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 4]
```diff
@@ -247,6 +247,57 @@ def test_to_datetime_parse_timezone_keeps_name(self):
...
+        if locale.getlocale() != ('zh_CN', 'UTF-8') and \
+           locale.getlocale() != ('it_IT', 'UTF-8'):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: dont use ``\`` on the break, use parens instead.

name this test: test_error_iso_week_year

**Review by jreback — CHANGES_REQUESTED** [review]
loooks pretty good. ping when updated and green.

**Comment by RjLi13** [comment]
@jreback I addressed comments and rebased my branch to latest.

File: pandas/_libs/tslibs/strptime.pyx [hunk 5]
```diff
@@ -597,7 +636,23 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
                               int day_of_week, int week_starts_Mon):
     """Calculate the Julian day based on the year, week of the year, and day of
     the week, with week_start_day representing whether the week of the year
-    assumes the week starts on Sunday or Monday (6 or 0)."""
+    assumes the week starts on Sunday or Monday (6 or 0).
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: can you type the return here (cdef int)

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 6]
```diff
@@ -247,6 +247,64 @@ def test_to_datetime_parse_timezone_keeps_name(self):
...
+        ["ISO week directive '%V' must be used with the ISO year directive "
+         "'%G' and a weekday directive '%A', '%a', '%w', or '%u'.", "20", "%V"]
+    ])
+    def test_error_iso_week_year(self, msg, s, _format):
+        # See GH#16607
+        # This test checks for errors thrown when giving the wrong format
+        # However, as discussed on PR#25541, overriding the locale
+        # causes a different error to be thrown due to the format being
+        # locale specific, but the test data is in english.
+        # Thus, a hack I did was to only run this test if locale was not
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: you can remove from here down (the comments) as it is enough by here

File: doc/source/whatsnew/v0.25.0.rst [hunk 7]
```diff
@@ -134,7 +134,7 @@ Categorical
 Datetimelike
 ^^^^^^^^^^^^
 
--
+- Added support for ISO week year format ('%G-%V-%u') when parsing datetimes using :meth: `to_datetime` (:issue:`16607`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: can you move this to Other Enhancements

**Review by jreback — CHANGES_REQUESTED** [review]
@RjLi13 lgtm. just some small comments. I think we need to put a link to the stftime docs in pd.to_datetime doc-string (like we do in doc/source/user_guide/timeseries.rst). can do here or in a followup.

File: pandas/_libs/tslibs/strptime.pyx (hunk 5 above)
Comment (RjLi13) [reply pandas/_libs/tslibs/strptime.pyx]: Sorry I don't understand what you mean. In the description of the summary or something else?

Comment (jreback) [reply pandas/_libs/tslibs/strptime.pyx]: no the return type of the  function (guess it wasn't typed before)

```
cdef int _calc_julian_from_U_or_W(......)
```

also you can type the one right below as ``cdef object .....``

**Comment by RjLi13** [comment]
RR (Review Ready)

**Comment by RjLi13** [comment]
@jreback @gfyoung @WillAyd could you continue your reviews before I rebase to latest master?

File: pandas/tests/indexes/datetimes/test_tools.py (hunk 6 above)
Comment (gfyoung) [reply pandas/tests/indexes/datetimes/test_tools.py]: @RjLi13 : Did you address this comment here?

Comment (RjLi13) [reply pandas/tests/indexes/datetimes/test_tools.py]: @gfyoung I believe I did. The comment was much longer but I took @jreback 's comment to mean everything below the line he commented and deleted it.

Comment (gfyoung) [reply pandas/tests/indexes/datetimes/test_tools.py]: Is the sentence starting with: "Thus, a hack I did was..." necessary?  Comments regarding first-person events are generally omitted, or should be reworded to not use first-person.

Comment (RjLi13) [reply pandas/tests/indexes/datetimes/test_tools.py]: @gfyoung I can refactor to not use first person. If you have something in mind already let me know

Comment (gfyoung) [reply pandas/tests/indexes/datetimes/test_tools.py]: If you could refactor to not use first person, that probably would be sufficient.

Comment (RjLi13) [reply pandas/tests/indexes/datetimes/test_tools.py]: Done

**Comment by RjLi13** [comment]
@jreback @gfyoung @WillAyd RR

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @RjLi13 really nice patch! keep em coming!

**Pull request merged**